### ISSUE TITLE
Provide a sandbox instance on smw.org

### ISSUE THREAD

**Issue by mwjames** [SemanticMediaWiki/SemanticMediaWiki#729]
Following-up on [0], it should be rather simple (given the Composer install) to provide a `https://sandbox.semantic-mediawiki.org/` instance for users to play and test without much cost (from a technical and human effort perspective).

@kghbln For spam protection maybe Karsten can weigh in on what needs to be done.

I prepared a [XML](https://gist.github.com/mwjames/2f83d3d775cbcc42b4ed) that can be used (via `Special:Import` or `importDump.php`) to import and setup some examples and/or for resetting the sandbox to an initial state.

[0] http://wikimedia.7.x6.nabble.com/Defining-duplicate-property-names-in-subobjects-tp5042657p5042674.html

**Comment by JeroenDeDauw** [comment]
This needs to be discussed on the OSDA list. @mwjames can you send emails to this thing? Perhaps not if you are not on it. In can always send one and cc you.

**Comment by mwjames** [comment]
> This needs to be discussed on the OSDA list.

Not on the list.

> @mwjames can you send emails to this thing? Perhaps not if you are not on it. In can always send one and cc you.

I hate administration, if someone can link to this issue I would be much obliged.

**Comment by JeroenDeDauw** [comment]
Apparently `https://sandbox.semantic-mediawiki.org/` existed in the past. Not sure what happened to it. Ping @kghbln

**Comment by JeroenDeDauw** [comment]
I asked @yaronkoren if referata (and thus http://scratchpad.referata.com/) would likely be updated in the next month or two, and he said probably not.

**Comment by kghbln** [comment]
``https://sandbox.semantic-mediawiki.org/`` was abandoned since it was unmaintained and spammed. At that time we still lived with the pre Composer 1.8 world and figured that `` http://scratchpad.referata.com/`` is the better and far more popular alternative. One reason for sandbox.smw.o not being maintained was that it was still "handled" via KIT and not OSDA. Now it redirects to semweb.o. which was not really the best choice I guess.

Basically what is needed is the one in control of the DNS for sandbox point it to the OSDA server and do a MW install + the Composer stuff for SMW. @mkroetzsch has the best insight here I suppose.

**Comment by JeroenDeDauw** [comment]
> redirects to semweb.o. which was not really the best choice I guess.

That wiki is down?!

> the one in control of the DNS for sandbox point it to the OSDA server

Is it not OSDA that owns the domain? In that case we ought to have DNS control over all subdomains no?

**Comment by kghbln** [comment]
> That wiki is down?!

Not a couple of day ago.

> Is it not OSDA that owns the domain? In that case we ought to have DNS control over all subdomains no?

This was my way of adding a ray of hope to the discussion that it probably will not take too long to get this act together.

**Comment by jongfeli** [comment]
@kghbln if you need help to setup a sandbox wiki from scratch I am willing
to help out. This is what a lot of people are missing and can be of great
help for debugging SMW.
Op 23 jan. 2015 10:23 schreef "Karsten Hoffmeyer" <notifications@github.com

**Comment by kghbln** [comment]
@jongfeli Thank you for offering you help. That's great. If there are on-wiki things to do you may definitively do so. As a matter of fact you are very welcome to do so. Installing the wiki as such requires access to the server so @JeroenDeDauw will be our man. :) Once this is done we see that @mwjames xml gets into the wiki.

Besides, after a lot of reluctance I started upgrading my farm from 1.8 to 2.1 today and this was much less painful than I originally expected as a matter of fact no pain at all was involved so far. So I believe I currently offer the only public SMW test wiki as of today. "Sadly" it is a German language instance.

@mwjames Is there a reason why we recommend people killing their semantic store prior to upgrading from 1.8? I do not really see a reason to do so.

**Comment by mwjames** [comment]
> @mwjames Is there a reason why we recommend people killing their semantic store prior to upgrading from 1.8? I do not really see a reason to do so and I have not done so without probs.

Did we, if so I can't recall the reason.

**Comment by mkroetzsch** [comment]
I have now deleted the subdomain DNS entry for sandbox.semantic-mediawiki.org, which means it will point to out main server (after DNS has updated everywhere). If somebody wanted to configure a sandbox wiki there, it would be possible now. I can also point to any other server if we had a sandbox elsewhere, of course.

**Comment by mwjames** [comment]
@JeroenDeDauw @kghbln It would be really neat if this could be resolved by the time of SMWCon (or during).

**Comment by kghbln** [comment]
I wonder why you got this idea today. ;) pass-throug ping to @JeroenDeDauw

**Comment by JeroenDeDauw** [comment]
While I like to see such a sandbox happen, I think it is a bad idea for me to take up this task. Installing and maintaining wikis is not something I like doing or am particularly good at. Just look at how frequently the main SMW wiki is getting updated... Perhaps someone who has a decent server wants to volunteer to run a sandbox wiki, and then we can point the DNS for the sandbox subdomain there?

**Comment by kghbln** [comment]
I really would not like to have access to the osda-server, however I was already thinking of providing a test instance on my private test server. Dunno about the load I will be getting, but we could at least test if this is a way to go.

**Comment by JeroenDeDauw** [comment]
We could ask around on the mailing list. You are already doing a pile of stuff in other areas

**Comment by kghbln** [comment]
Ah, I was more thinking about the server load from the traffic. I have 1 GB of RAM and half a CPU (MHz: 2266.746, L1d cache: 32K, L1i cache: 32K, L2 cache: 4096K)

**Comment by mwjames** [comment]
> We could ask around on the mailing list. You are already doing a pile of stuff in other areas

This is going into the wrong direction. If we cannot make a sandbox then we just let people use smw.org to test things instead of pushing them off or link them to an outdated release.

People want to try things out before they start to install the software, so let them.

**Comment by kghbln** [comment]
My idea or better understanding of this sandbox is that it always runs latest stable (not master) of MW as well as SMW and (semantic) extensions. Apart from that I'd rather like more active testers than somebody agreeing to provide a sandbox.

To cut it short. I will try to set it up this weekend so that the DNS may be changed to the respective IP.

**Comment by JeroenDeDauw** [comment]
@mwjames I think you do not understand what I was getting at. My suggestion is not to point people to some random badly maintained wiki out there where they can test stuff. Rather, I'd like to see a test wiki that is actively maintained, so it does not get out of date, and so that new things that people want to test can be enabled in timely manner. Sure we can do this ourselves, though that takes our time away from other SMW things we are already doing. Having a new dedicated volunteer to take care of this would be great no?

**Comment by mwjames** [comment]
>  Having a new dedicated volunteer to take care of this would be great no?

It would be great in an ideal world and now and then I'm surprised by a light of hope that shines through the cracks (as with #1165) but I have come to terms with the fact that nothing is happing if you are not pushing for it, contribution is limited to the existing candidates which you can count fairly easily.

While I would welcome more volunteers and have users scream "I will do it, I'll help you ...", the hard reality shows its ugly face (and that is what experience dictates not the mere hope for an utopia).

**Comment by kghbln** [comment]
So this is done now. Just add IP 217.197.83.171 to the A-Record of "sandbox.sematic-mediawiki.org".

**Comment by mwjames** [comment]
> So this is done now. 

Well done.

![image](https://cloud.githubusercontent.com/assets/1245473/10119242/33a92996-6490-11e5-902f-5430b013cab9.png)

PS: The super contributor badge was not available so I took the next best thing.

**Comment by mwjames** [comment]
@kghbln Did I miss something?

![image](https://cloud.githubusercontent.com/assets/1245473/10119250/9fca5866-6490-11e5-8472-46b18b868f93.png)

**Comment by kghbln** [comment]
This is the default page of the testserver. First the DNS configuration has to be updated for it to work. A little patience is needed until @JeroenDeDauw worked on this.

**Comment by kghbln** [comment]
>Well done.

Cooool. :D 

Currently "sandbox.semantic-mediawiki.org" redirects to "semantic-mediawiki.org"

**Comment by mkroetzsch** [comment]
Great, many thanks! I have now changed the A-entry for sandbox.semantic-mediawiki.org. Works for me, but it might still take a moment until all DNS serves have it.

**Comment by JeroenDeDauw** [comment]
\o/ Thanks @kghbln, @mwjames and @mkroetzsch

**Comment by JeroenDeDauw** [comment]
> It would be great in an ideal world and now and then I'm surprised by a light of hope that shines through the cracks (as with #1165) but I have come to terms with the fact that nothing is happing if you are not pushing for it, contribution is limited to the existing candidates which you can count fairly easily.

> While I would welcome more volunteers and have users scream "I will do it, I'll help you ...", the hard reality shows its ugly face (and that is what experience dictates not the mere hope for an utopia). 

Reality is indeed quite different than a stampede of people that dedicate all their free time to help out without first being further incentivized. Most people will not contribute back. Some will, though most of these need a nudge to get started, or at the very least some easy thing to start with. If we do not clearly communicate contributions of various types are welcome (in a way understandable to the people that are potentially interested in contributing), and we do not have easy tasks for those people, then indeed, it is silly to expect new contributors to appear. This is how I got involved via GSoC. It's how @jongfeli got started with submitting CS improvements. It's how several people got into the testers group and are now helping out with that. I think you'll find there are a multiple awesome people like @kghbln in our community that'd be willing to set up a wiki such as discussed here if you ask.

**Comment by kghbln** [comment]
I dunno where I read it but someone suggested the sandbox to run on master when it comes to the semantic extensions located at GitHub. I can do this since it is just a matter of changing the "composer.local.json" file. I have create a cronjob to do the rest every night anyways.

**Comment by kghbln** [comment]
This was on twitter by @JeroenDeDauw. Since @mwjames also indicated via e-mail that master would be perferable, well, it's done. :)

**Comment by JeroenDeDauw** [comment]
Is master really preferable? I'd expect the main use case for the sandbox to be that people can just play around with stuff that they can also use on their wiki. For that stable releases are nice. Running master might make the wiki go down some of the time, as not everything is always compatible with each other. (Also depends on which things you run master for... can't do this for all, as composer will say no.jpg. You can do dev versions within the compatible ranges of course.) Having the latest possible dev versions show up is nice for developers and pre release testers. That is quite a different use case though. Given these two downsides for the use case I'm assuming is the main one, I'd not combine these two.

**Comment by mwjames** [comment]
My intake here is that smw.org runs the stable version with examples provided but sandbox.semantic-mediawiki.org runs the dev version which in turn can be used to compare results or allows for quicker intervention.

I do believe that ours tests (at least for SMW core) guard us against any big-boom-cannot-be-installed scenario. Of all extensions on one box makes verification a bit easier when someone comes and says "it doesn't work". 

> test.semantic-mediawiki.org - dev

I think a bit too much infrastructure, let's keep lean and clean.

**Comment by JeroenDeDauw** [comment]
> My intake here is that smw.org runs the stable version with examples provided but sandbox.semantic-mediawiki.org runs the dev version which in turn can be used to compare results or allows for quicker intervention.

The main SMW wiki is not intended to be used as a sandbox. Also, the concept of a sandbox is disjoint from running development versions or testing those.

> I do believe that ours tests (at least for SMW core) guard us against any big-boom-cannot-be-installed scenario. Having all extensions in one box makes verification a bit easier when someone comes and says "it doesn't work". 

For SMW yes. But not for all extensions. If you run SMW master, and there is a breaking change there, it will be incompatible with extensions that relied on the old behaviour, until fixes are made there. This is why the composer dependencies have upper bounds in their version ranges and don't just specify master.

> I think a bit too much infrastructure, let's keep lean and clean.

You are right in that this is an additional wiki. And I'm feeling a bit bad for asking for another one to be created. However, I do not agree this is less lean and clean. Single Responsibility Principle. Mashing different concerns into a single class is not more clean, same goes for infrastructure :)

**Comment by kghbln** [comment]
Until 2.3 was released sandbox should stay on master for the simple reason that already Semantic Cite specific content was added. We could then switch to stable as I provide another instance running master. I would call the subdomain "testmaster" since "sandbox" and "test" cannot really be distinguished intuitively by occasional users. Doing it later will also allow me to have more insight into what amount of traffic the current sandbox is getting. This is currently not a very strong VM.

Besides on my private wiki running master I had three fatals within a year, two caused by SMW and one caused by a wmf maintained extension dropping compat for previous versions of MW (well one should not do this anyway if you do not intend to run MW on master also). So there are luckly just a few worries.

**Comment by mwjames** [comment]
>I had three fatals within a year, two caused by SMW and one caused by a wmf maintained extension dropping compat for previous versions of MW

Speaking of the devil ...

![image](https://cloud.githubusercontent.com/assets/1245473/10216679/25e81102-682b-11e5-91b3-5e3ab976e97f.png)

**Comment by kghbln** [comment]
Hmm, it's working for me. We should hover open an extra issue for reporting troubles with sandbox

**Comment by mwjames** [comment]
> We should hover open an extra issue for reporting troubles with sandbox
> of MW (well one should not do this anyway if you do not intend to run MW on master also).

Would be wise I just encountered:
```
A database query error has occurred. This may indicate a bug in the software.

    Query:
    SELECT 1 FROM `02100_page` WHERE (Array) OR (Array) OR (Array) LIMIT 1
    Function: ApiStashEdit::checkCache
    Error: 1054 Unknown column 'Array' in 'where clause' (localhost)
```

Caused most likely by https://phabricator.wikimedia.org/T105597

**Comment by kghbln** [comment]
I encountered this issue about two months ago after @thingles updated WikiApiary (https://github.com/WikiApiary/wiki/issues/4) This is quite nasty. I have however not configured sandbox to use redis yet.

**Comment by kghbln** [comment]
The tracking issue for "sandbox.semantic-mediawiki.org" is now at https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/1186
